solc/CommandLineInterface.cpp

```diff
@@ -916,67 +918,134 @@ void CommandLineInterface::serveLSP()
+
 	for (auto& src: sourceCodes)
 	{
-		auto end = src.second.end();
-		for (auto it = src.second.begin(); it != end;)
+		// Map from how the libraries will be named inside the bytecode to their addresses.
+		map<size_t, string> linkReferences;
+		string bytecode;
+		vector<string> sourceLines;
+		boost::split(sourceLines, src.second, boost::is_any_of("\n"));
+
+		size_t bytecodeIndex;
+		bool bytecodeFound = false;
+		for (size_t i = 0; i < sourceLines.size(); ++i)
 		{
-			while (it != end && *it != '_') ++it;
-			if (it == end) break;
-			if (
-				end - it < placeholderSize ||
-				*(it + 1) != '_' ||
-				*(it + placeholderSize - 2) != '_' ||
-				*(it + placeholderSize - 1) != '_'
-			)
+			// If all characters are bytecode characters
+			if (sourceLines[i].size() > 0
+				&& all_of(sourceLines[i].begin(), sourceLines[i].end(), isBytecodeChar))
+			{
+				bytecodeFound = true;
+				bytecodeIndex = i;
+				break;
+			}
+		}
+
+		if (!bytecodeFound)
+			continue;
+
+		for (auto const& library: libraries)
+		{
+			string const& name = library.first;
+			// Bytecode we need is on the third line there's a blank line at the start
+			bytecode = sourceLines.at(bytecodeIndex);
+			size_t libraryStartOffset = bytecode.find("__$");
+			size_t libraryEndOffset = bytecode.find("$__");
+
+			bool startMarkerFound = libraryStartOffset != string::npos;
+			bool endMarkerFound = libraryEndOffset != string::npos;
+
+			if (!startMarkerFound && !endMarkerFound)
+			{
+				// No placeholders found
+				continue;
+			}
+
+			if (startMarkerFound && !endMarkerFound)
+			{
 				solThrow(
 					CommandLineExecutionError,
-					"Error in binary object file " + src.first + " at position " + to_string(it - src.second.begin()) + "\n" +
-					'"' + string(it, it + min(placeholderSize, static_cast<int>(end - it))) + "\" is not a valid link reference."
+					"Error in binary object file " + src.first + " unbounded placeholder at " + to_string(libraryStartOffset) + "\n"
 				);
+			}
 
-			string foundPlaceholder(it, it + placeholderSize);
-			if (librariesReplacements.count(foundPlaceholder))
+			if (!startMarkerFound && endMarkerFound)
 			{
-				string hexStr(toHex(librariesReplacements.at(foundPlaceholder).asBytes()));
-				copy(hexStr.begin(), hexStr.end(), it);
+				solThrow(
+					CommandLineExecutionError,
+					"Error in binary object file " + src.first + " unbounded placeholder without start point __$ at " + to_string(libraryEndOffset) + "\n"
+							);
+			}
+
+			if (startMarkerFound && endMarkerFound)
+			{
+				if ((libraryEndOffset - libraryStartOffset) != placeholderSize - 3)
+				{
+					solThrow(
+						CommandLineExecutionError,
+						"Error in binary object file " + src.first + " truncated placeholder found at " + to_string(libraryStartOffset) +
+							" placeholder addresses should be " + to_string(placeholderSize) + " characters long (including address markers __$ and $__)\n"
+					);
+				}
+			}
```

Comment:
You don't need braces if there's just one instruction inside `if` or `for`. The code is more concise without them:
```suggestion
			if (startMarkerFound && endMarkerFound)
				if ((libraryEndOffset - libraryStartOffset) != placeholderSize - 3)
					solThrow(
						CommandLineExecutionError,
						"Error in binary object file " + src.first + " truncated placeholder found at " + to_string(libraryStartOffset) +
							" placeholder addresses should be " + to_string(placeholderSize) + " characters long (including address markers __$ and $__)\n"
					);
```
There are also a few other places like this in the PR.